flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/FailoverRegion.java

```diff
@@ -0,0 +1,251 @@
+		}
+	}
+
+	public JobStatus getState() {
+		return state;
+	}
+
+	/**
+	 * get all execution vertexes contained in this region
+	 */
+	public List<ExecutionVertex> getAllExecutionVertexes() {
+		return connectedExecutionVertexes;
+	}
+
+	// Notice the region to failover, 
+	private void failover(long globalModVersionOfFailover) {
+		if (!executionGraph.getRestartStrategy().canRestart()) {
+			executionGraph.failGlobal(new FlinkException("RestartStrategy validate fail"));
+		}
+		else {
+			JobStatus curStatus = this.state;
+			if (curStatus.equals(JobStatus.RUNNING)) {
+				cancel(globalModVersionOfFailover);
+			}
+			else if (curStatus.equals(JobStatus.CANCELED)) {
+				reset(globalModVersionOfFailover);
+			}
+			else {
+				LOG.info("FailoverRegion {} is {} when notified to failover.", id, state);
+			}
+		}
+	}
+
+	// cancel all executions in this sub graph
+	private void cancel(final long globalModVersionOfFailover) {
+		while (true) {
+			JobStatus curStatus = this.state;
+			if (curStatus.equals(JobStatus.RUNNING)) {
+				if (transitionState(curStatus, JobStatus.CANCELLING)) {
+
+					// we build a future that is complete once all vertices have reached a terminal state
+					final ArrayList<Future<?>> futures = new ArrayList<>(connectedExecutionVertexes.size());
+
+					// cancel all tasks (that still need cancelling)
+					for (ExecutionVertex vertex : connectedExecutionVertexes) {
+						futures.add(vertex.cancel());
+					}
+
+					final FutureUtils.ConjunctFuture allTerminal = FutureUtils.combineAll(futures);
+					allTerminal.thenAcceptAsync(new AcceptFunction<Void>() {
+						@Override
+						public void accept(Void value) {
+							allVerticesInTerminalState(globalModVersionOfFailover);
+						}
+					}, executor);
+
+					break;
+				}
+			}
+			else {
+				LOG.info("FailoverRegion {} is {} when cancel.", id, state);
+				break;
+			}
+		}
+	}
+
+	// reset all executions in this sub graph
+	private void reset(long globalModVersionOfFailover) {
+		try {
+			// reset all connected ExecutionVertexes
+			final Collection<CoLocationGroup> colGroups = new HashSet<>();
+			final long restartTimestamp = System.currentTimeMillis();
+
+			for (ExecutionVertex ev : connectedExecutionVertexes) {
+				CoLocationGroup cgroup = ev.getJobVertex().getCoLocationGroup();
+				if (cgroup != null && !colGroups.contains(cgroup)){
+					cgroup.resetConstraints();
+					colGroups.add(cgroup);
+				}
+
+				ev.resetForNewExecution(restartTimestamp, globalModVersionOfFailover);
+			}
+			if (transitionState(JobStatus.CANCELED, JobStatus.CREATED)) {
+				restart(globalModVersionOfFailover);
+			}
+			else {
+				LOG.info("FailoverRegion {} switched from CANCELLING to CREATED fail, will fail this region again.", id);
+				failover(globalModVersionOfFailover);
+			}
+		}
+		catch (GlobalModVersionMismatch e) {
+			// happens when a global recovery happens concurrently to the regional recovery
+			// go back to a clean state
+			state = JobStatus.CREATED;
```

Comment:
The initial state of FailoverRegion should be RUNNING, as when FailoverRegion is working, the exeuctions in it will always be scheduled already. The CREATED state is only used during the failover process.